activerecord/test/cases/touch_later_test.rb

```diff
@@ -120,4 +121,101 @@ def test_touching_three_deep
     assert_not_equal nodes(:grandparent).reload.updated_at, previous_grandparent_updated_at
     assert_not_equal trees(:root).reload.updated_at, previous_tree_updated_at
   end
+
+  class FullCommitTest < ActiveRecord::TestCase
+    self.use_transactional_tests = false
+
+    fixtures :nodes, :trees
+
+    def self.run(*args)
+      super unless current_adapter?(:SQLite3Adapter)
+    end
+
+    module WaitForConcurrentCommit
+      attr_accessor :concurrent_thread
+
+      def before_committed!
+        super
+        concurrent_thread&.wait(1)
+        self.concurrent_thread = nil
+      end
+    end
+
+    class ConcurrentTree < ActiveRecord::Base
+      prepend WaitForConcurrentCommit
+      self.table_name = "trees"
+    end
+
+    class ConcurrentNode < ActiveRecord::Base
+      prepend WaitForConcurrentCommit
+      self.table_name = "nodes"
+
+      belongs_to :tree, class_name: ConcurrentTree.name, touch: true
+      belongs_to :parent, class_name: self.name, touch: true
```

Comment:
The deadlock in `test_touch_callback_deadlocks_by_association_depth` is caused by the order in which these associations are declared: updating a second-level node touches the tree first (moving up two levels in the association hierarchy) and then its parent node (moving down a level again). I don't think that Rails should reorder these touches automatically, given that the developer can fix the issue by changing the order of the associations.

More generally: as long as the application ensures that touches only cascade in one direction, we can avoid the problem described in https://github.com/rails/rails/pull/38146#issuecomment-570694381 without reflecting on the model's associations by preserving the order that tables first appear in the `records` array:

```patch
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb b/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
index 7893ec85d6..aea1bd8371 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
@@ -116,7 +116,13 @@ def rollback_records
       end

       def before_commit_records
-        records.uniq.each(&:before_committed!) if records && @run_commit_callbacks
+        return unless records && @run_commit_callbacks
+
+        records.uniq.group_by do |record|
+          record.class.table_name
+        end.each_value do |records|
+          records.sort_by(&:id).each(&:before_committed!)
+        end
       end

       def commit_records
```

Reply:
Yes, the declaration order within one model is obviously one way of causing this. But assuming the author will need to take that into account seems less "declarative" and much harder to trace and prove correct between different models.

Something very general like `belongs_to :user` (depth=1) is most typically going to be the first thing declared on several common models above other associations (that themselves may indirectly belong to other users).

> as long as the application ensures that touches only cascade in one direction

That's a very reasonable assumption and right now I only perform the depth lookups for `belongs_to` as well. But thinking about this, there's no reason we couldn't also consider `has_one` depths for completeness. (Still, I would assume those are edge-cases in any model tree. Typically the deepest upward traversal would be via `belongs_to`.)

I did consider using `group_by` to sort by the first occurence of each table instead of name order. However, that doesn't promise the order will be the same between different transaction blocks.

It very much depends on the data and runtime conditions, too. Just for one example, associations can be optional:

Transaction A:

```
Child.create(parent_id: nil, school_id: 1)
Child.create(parent_id: 1, school_id: nil)

transaction.records == [#<School id: 1>, #<Parent id: 1>]
```

Transaction B:

```
Child.create(parent_id: 1, school_id: nil)
Child.create(parent_id: nil, school_id: 1)

transaction.records == [#<Parent id: 1>, #<School id: 1>]
```

I would still recommend that ActiveRecord should apply some constant, deterministic ordering to these callbacks.

Reply:
@eugeneius I could rewrite the test with the associations declared in your suggested "correct" order, and provide additional scenarios for the other examples I showed here.

I don't think the `group_by` sorting is sufficient so maybe adding a test case to fail that implementation would be helpful.

Please let me know how I can help move this forward.

Reply:
@eugeneius is there still any interest in this patch?

Let me know so I can update this.